packages/razzle-plugin-scss/tests/index.test.js

```diff
@@ -0,0 +1,113 @@
+'use strict';
+
+const createConfig = require('razzle/config/createConfig');
+const {
+  cssLoaderFinder,
+  postCssLoaderFinder,
+  resolveUrlLoaderFinder,
+  sassLoaderFinder,
+  styleLoaderFinder,
+} = require('../helpers');
+const pluginFunc = require('../index');
+
+describe('razzle-scss-plugin', () => {
+  describe('when creating a web config', () => {
+    describe('when environment set to development', () => {
+      let config;
+      beforeAll(() => {
+        config = createConfig('web', 'dev', {
+          plugins: [{ func: pluginFunc }],
+        });
+      });
+
+      it('should add style-loader', () => {
+        const rule = config.module.rules.find(styleLoaderFinder);
+        expect(rule).not.toBeUndefined();
```

Comment:
When I have a lot of tests that have exactly the same structure, I prefer to just keep them in an array and just call forEach on them. Helps stay DRY.  

For example:

```js
const devLoaderTests = [{
   name: '...',
   loaderFinder: '...'
}, {
 // ...
},
 {
 // ...
}]


describe('dev', () => {
  devLoaderTests.forEach(item => {
    it(item.name, () => {
     expect(config.module.rules.find(item.loaderFinder).not.toBeUndefined());
    })
  });
});
```

Reply:
Ahh sorry for repeating codes. I will fix it!

Reply:
here's what I do in mocha for recursive:

```js
  const template = (strings, ...keys) => (...values) => {
    const dict = values[values.length - 1] || {};
    const result = [strings[0]];
    keys.forEach((key, i) => {
      const value = Number.isInteger(key) ? values[key] : dict[key];
      result.push(value, strings[i + 1]);
    });
    return result
      .join("")
      .replace(/\s+/g, " ")
      .trim();
  },

  const generateTest = (authenticated, credential, templates, tests) => {
    return _.zipWith(templates, tests, (template, test) => {
      if (_.isFunction(template)) {
        it(template({ ...test.props, ...credential }), async () => {
          if (_.isObject(test)) {
            if (_.isUndefined(test.payload)) {
              test.check(
                await authenticated[test.method](
                  module.exports.baseUrl(test.url)
                )
              );
            } else {
              test.check(
                await authenticated[test.method](
                  module.exports.baseUrl(test.url)
                ).send(test.payload)
              );
            }
          }
        });
      } else {
        describe(template.desc({ role: credential.role }), () => {
          return module.exports.generateTest(
            authenticated,
            credential,
            template.test,
            test
          );
        });
      }
    });
```

```js
const templateTest = [
  template`should ${"should"} get list of users if ${"role"} is logged in`,
  template`should ${"should"} edit users if ${"role"} is logged in`,
  {
    desc: template`details of specific user by ${"role"}`,
    test: [
      template`should ${"should"} get details of specific wrong id of user because resource not found`,
      template`should ${"should"} get details of specific correct id of user`
    ]
  }
];
```

```js
const authenticatedTest = [{
   credentials: {
    username: "administrator",
    password: "administrator",
    role: "admin"
   },
   tests: [{
     url: "/users",
     method: `get`,
     check: result => {
      expect(result).to.have.status(200);
      expect(result.body).to.have.ownProperty("data");
      expect(result.body.data).to.be.an("array");
      expect(result.body).to.have.all.keys(paginateProps);
     }
    }, {
     props: {
      should: "not"
     },
     url: `/users/${user.id_user}`,
     payload: user,
     method: `patch`,
     check: result => {
      expect(result).to.have.status(403);
      expect(result.body.message).to.equal("Forbidden");
      expect(result.body.name).to.equal("ForbiddenError");
     }
    },
    [{
     props: {
      should: "not"
     },
     url: `/user/${user.wrong_id_user}`,
     method: `get`,
     check: result => {
      expect(result).to.have.status(404);
      expect(result.body.message).to.equal("User not found.");
      expect(result.body.name).to.equal("NotFoundError");
     }
    }, {
     url: `/user/${user.id_user}`,
     method: `get`,
     check: result => {
      expect(result).to.have.status(200);
      expect(result.body).to.be.an("object");
      expect(result.body).to.include.all.keys(userColumns);
     }
    }],
    ...
   ];
```

and then generate it like this

```js
 authenticatedTest.forEach(auth => {
   describe(`authenticated role : ${auth.credentials.role}`, () => {
    const authenticated = chai.request.agent(routes);

    before(() =>
     knex.migrate
     .rollback()
     .then(() => knex.migrate.latest())
     .then(() => knex.seed.run())
     .then(() =>
      authenticated
      .post(`${config.get("routePrefix")}/session`)
      .send(auth.credentials)
     )
     .then(res => {
      expect(res).to.have.status(200);
      expect(res.body).to.have.ownProperty("username");
      expect(res.body).to.have.ownProperty("status");
     })
    );

    after(() =>
     authenticated.del(`${config.get("routePrefix")}/session`).then(res => {
      expect(res).to.have.status(200);
      expect(res.body).to.be.an("object");
      expect(res.body.message).to.equal("Logged out successfully.");
     })
    );

    generateTest(authenticated, auth.credentials, templateTest, auth.tests);
  });
```

what do you think?